tests/test_cmd2.py

```diff
@@ -1388,7 +1389,106 @@ def test_echo(capsys):
     assert app._current_script_dir is None
     assert out.startswith('{}{}\n'.format(app.prompt, command) + 'history [arg]: lists past commands issued')
 
-
+def test_pseudo_raw_input_tty_rawinput_true():
+    # use context managers so original functions get put back when we are done
+    # we dont use decorators because we need m_input for the assertion
+    with mock.patch('sys.stdin.isatty',
+            mock.MagicMock(name='isatty', return_value=True)):
+        with mock.patch('six.moves.input',
+                mock.MagicMock(name='input', side_effect=['set', EOFError])) as m_input:
+            # run the cmdloop, which should pull input from our mocks
+            app = cmd2.Cmd()
+            app.use_rawinput = True
+            app._cmdloop()
+            # because we mocked the input() call, we won't get the prompt
+            # or the name of the command in the output, so we can't check
+            # if its there. We assume that if input got called twice, once
+            # for the 'set' command, and once for the 'quit' command,
+            # that the rest of it worked
+            assert m_input.call_count == 2
+    
+def test_pseudo_raw_input_tty_rawinput_false():
+    # gin up some input like it's coming from a tty
+    fakein = io.StringIO(u'{}'.format('set\n'))
+    mtty = mock.MagicMock(name='isatty', return_value=True)
+    fakein.isatty = mtty
+    mreadline = mock.MagicMock(name='readline', wraps=fakein.readline)
+    fakein.readline = mreadline
+    
+    # run the cmdloop, telling it where to get input from
+    app = cmd2.Cmd(stdin=fakein)
+    app.use_rawinput = False
+    app._cmdloop()
+    
+    # because we mocked the readline() call, we won't get the prompt
+    # or the name of the command in the output, so we can't check
+    # if its there. We assume that if readline() got called twice, once
+    # for the 'set' command, and once for the 'quit' command,
+    # that the rest of it worked
+    assert mreadline.call_count == 2
+
+# the next helper function and two tests check for piped
+# input when use_rawinput is True.
+def piped_rawinput_true(capsys, echo, command):
+    app = cmd2.Cmd()
+    app.use_rawinput = True
+    app.echo = echo
+    # run the cmdloop, which should pull input from our mock
+    app._cmdloop()
+    out, err = capsys.readouterr()
+    return (app, out)
+
+# using the decorator puts the original function at six.moves.input
+# back when this method returns
+@mock.patch('six.moves.input',
```

Comment:
I wasn't aware you could use mocks as decorators either.  This is another great feature.